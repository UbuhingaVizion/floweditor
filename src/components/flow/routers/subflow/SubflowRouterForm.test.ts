import SubflowRouterForm from 'components/flow/routers/subflow/SubflowRouterForm';
import { FlowTypes } from 'config/interfaces';
import { composeComponentTestUtils, mock, setMock } from 'testUtils';
import {
  ColorFlowAsset,
  SoundFlowAsset,
  createStartFlowAction,
  createVoiceStartFlowAction,
  createSubflowNode,
  getRouterFormProps
} from 'testUtils/assetCreators';
import * as utils from 'utils';

const msgRouterNode = createSubflowNode(createStartFlowAction(), null, FlowTypes.MESSAGING);
const { setup: msgSetup } = composeComponentTestUtils(
  SubflowRouterForm,
  getRouterFormProps(msgRouterNode)
);
mock(utils, 'createUUID', utils.seededUUIDs());

describe(SubflowRouterForm.name, () => {
  describe('render', () => {
    it('should render', () => {
      const { wrapper } = msgSetup();
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('updates', () => {
    it('should update and save', () => {
      const { instance, props } = msgSetup(true, { updateRouter: setMock() });
      instance.handleFlowChanged([ColorFlowAsset]);
      expect(instance.state).toMatchSnapshot();
      instance.getButtons().primary.onClick();
      expect(props.updateRouter).toMatchCallSnapshot();
    });

    it('should cancel changes', () => {
      const { instance, props } = msgSetup(true, { updateRouter: setMock() });
      instance.handleFlowChanged([ColorFlowAsset]);
      instance.getButtons().secondary.onClick();
      expect(props.updateRouter).not.toBeCalled();
    });

    it('converts from other node types', () => {
      const { instance, props } = msgSetup(true, {
        updateRouter: setMock(),
        nodeSettings: {
          originalNode: { ui: { $merge: { type: null } } }
        }
      });

      instance.handleFlowChanged([ColorFlowAsset]);
      instance.getButtons().primary.onClick();
      expect(props.updateRouter).toMatchCallSnapshot();
    });

    it('creates its own action uuid if necessary', () => {
      const { instance, props } = msgSetup(true, {
        updateRouter: setMock(),
        nodeSettings: {
          originalNode: {
            node: { $merge: { actions: [] } },
            ui: { $merge: { type: null } }
          }
        }
      });

      instance.handleFlowChanged([ColorFlowAsset]);
      instance.getButtons().primary.onClick();
      expect(props.updateRouter).toMatchCallSnapshot();
    });

    it('validates before saving', () => {
      const { instance, props } = msgSetup(true, { updateRouter: setMock() });
      instance.handleFlowChanged([]);
      instance.getButtons().primary.onClick();
      expect(props.updateRouter).not.toBeCalled();
    });
  });
});

const voiceRouterNode = createSubflowNode(createVoiceStartFlowAction(), null, FlowTypes.VOICE);
const { setup: voiceSetup } = composeComponentTestUtils(
  SubflowRouterForm,
  getRouterFormProps(voiceRouterNode)
);
mock(utils, 'createUUID', utils.seededUUIDs());

describe(SubflowRouterForm.name, () => {
  describe('render', () => {
    it('should render', () => {
      const { wrapper } = voiceSetup();
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('updates', () => {
    it('should update and save', () => {
      const { instance, props } = voiceSetup(true, { updateRouter: setMock() });
      instance.handleFlowChanged([SoundFlowAsset]);
      expect(instance.state).toMatchSnapshot();
      instance.getButtons().primary.onClick();
      expect(props.updateRouter).toMatchCallSnapshot();
    });

    it('should cancel changes', () => {
      const { instance, props } = voiceSetup(true, { updateRouter: setMock() });
      instance.handleFlowChanged([SoundFlowAsset]);
      instance.getButtons().secondary.onClick();
      expect(props.updateRouter).not.toBeCalled();
    });

    it('converts from other node types', () => {
      const { instance, props } = voiceSetup(true, {
        updateRouter: setMock(),
        nodeSettings: {
          originalNode: { ui: { $merge: { type: null } } }
        }
      });

      instance.handleFlowChanged([SoundFlowAsset]);
      instance.getButtons().primary.onClick();
      expect(props.updateRouter).toMatchCallSnapshot();
    });

    it('creates its own action uuid if necessary', () => {
      const { instance, props } = voiceSetup(true, {
        updateRouter: setMock(),
        nodeSettings: {
          originalNode: {
            node: { $merge: { actions: [] } },
            ui: { $merge: { type: null } }
          }
        }
      });

      instance.handleFlowChanged([SoundFlowAsset]);
      instance.getButtons().primary.onClick();
      expect(props.updateRouter).toMatchCallSnapshot();
    });

    it('validates before saving', () => {
      const { instance, props } = voiceSetup(true, { updateRouter: setMock() });
      instance.handleFlowChanged([]);
      instance.getButtons().primary.onClick();
      expect(props.updateRouter).not.toBeCalled();
    });
  });
});
