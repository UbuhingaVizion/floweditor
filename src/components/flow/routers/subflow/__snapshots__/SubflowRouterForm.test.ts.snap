// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubflowRouterForm render should render 1`] = `
<Dialog
  buttons={
    Object {
      "primary": Object {
        "name": "Ok",
        "onClick": [Function],
      },
      "secondary": Object {
        "name": "Cancel",
        "onClick": [Function],
      },
    }
  }
  headerClass="enter_flow"
  tabs={Array []}
  title="Enter a Flow"
>
  <TypeList
    __className=""
    initialType={
      Object {
        "aliases": Array [
          "split_by_subflow",
        ],
        "component": [Function],
        "description": "Enter another flow",
        "form": [Function],
        "localization": [Function],
        "localizeableKeys": Array [
          "exits",
        ],
        "name": "Enter a Flow",
        "type": "enter_flow",
      }
    }
    onChange={[MockFunction]}
  />
  <AssetSelector
    assets={
      Object {
        "items": Object {},
        "type": "flow",
      }
    }
    entry={
      Object {
        "value": Object {
          "name": "Colors",
          "type": "messaging",
          "uuid": "b1f332f3-bdd3-4891-aec5-1843a712dbf1",
        },
      }
    }
    name="Flow"
    onChange={[Function]}
    placeholder="Select the flow to start"
    searchable={true}
    shouldExclude={[Function]}
  />
</Dialog>
`;

exports[`SubflowRouterForm render should render 2`] = `
<Dialog
  buttons={
    Object {
      "primary": Object {
        "name": "Ok",
        "onClick": [Function],
      },
      "secondary": Object {
        "name": "Cancel",
        "onClick": [Function],
      },
    }
  }
  headerClass="enter_flow"
  tabs={Array []}
  title="Enter a Flow"
>
  <TypeList
    __className=""
    initialType={
      Object {
        "aliases": Array [
          "split_by_subflow",
        ],
        "component": [Function],
        "description": "Enter another flow",
        "form": [Function],
        "localization": [Function],
        "localizeableKeys": Array [
          "exits",
        ],
        "name": "Enter a Flow",
        "type": "enter_flow",
      }
    }
    onChange={[MockFunction]}
  />
  <AssetSelector
    assets={
      Object {
        "items": Object {},
        "type": "flow",
      }
    }
    entry={
      Object {
        "value": Object {
          "name": "Sounds",
          "type": "voice",
          "uuid": "b1f332f3-bdd3-4891-aec5-1843a712dbf1",
        },
      }
    }
    name="Flow"
    onChange={[Function]}
    placeholder="Select the flow to start"
    searchable={true}
    shouldExclude={[Function]}
  />
</Dialog>
`;

exports[`SubflowRouterForm updates converts from other node types 1`] = `
Array [
  Object {
    "inboundConnections": Object {},
    "node": Object {
      "actions": Array [
        Object {
          "flow": Object {
            "name": "Favorite Color",
            "type": "messaging",
            "uuid": "9a93ede6-078f-44c9-ad0a-133793be5d56",
          },
          "type": "enter_flow",
          "uuid": "d4a3a01c-1dee-4324-b107-4ac7a21d836f",
        },
      ],
      "exits": Array [
        Object {
          "destination_uuid": null,
          "uuid": "b1f332f3-bdd3-4891-aec5-1843a712dbf1",
        },
        Object {
          "destination_uuid": null,
          "uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
        },
      ],
      "router": Object {
        "cases": Array [
          Object {
            "arguments": Array [
              "completed",
            ],
            "category_uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
            "type": "has_only_text",
            "uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
          },
          Object {
            "arguments": Array [
              "expired",
            ],
            "category_uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
            "type": "has_only_text",
            "uuid": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
          },
        ],
        "categories": Array [
          Object {
            "exit_uuid": "b1f332f3-bdd3-4891-aec5-1843a712dbf1",
            "name": "Complete",
            "uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
          },
          Object {
            "exit_uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
            "name": "Expired",
            "uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
          },
        ],
        "default_category_uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
        "operand": "@child.status",
        "type": "switch",
      },
      "uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
    },
    "ui": Object {
      "config": Object {},
      "position": null,
      "type": "split_by_subflow",
    },
  },
]
`;

exports[`SubflowRouterForm updates converts from other node types 2`] = `
Array [
  Object {
    "inboundConnections": Object {},
    "node": Object {
      "actions": Array [
        Object {
          "flow": Object {
            "name": "Favorite Sound",
            "type": "voice",
            "uuid": "80b0a850-6973-11ed-9c42-ad7d71773e93",
          },
          "type": "enter_flow",
          "uuid": "2d1284d0-6972-11ed-9ca7-8f291517c832",
        },
      ],
      "exits": Array [
        Object {
          "destination_uuid": null,
          "uuid": "a62e8284-49f8-4292-adbc-d5f6607f85c5",
        },
      ],
      "router": Object {
        "cases": Array [
          Object {
            "arguments": Array [
              "completed",
            ],
            "category_uuid": "0c8c9239-9b59-473a-a229-d8a9693be270",
            "type": "has_only_text",
            "uuid": "9adce29f-da95-4182-9cb3-0f1e3bf714e5",
          },
        ],
        "categories": Array [
          Object {
            "exit_uuid": "a62e8284-49f8-4292-adbc-d5f6607f85c5",
            "name": "Complete",
            "uuid": "0c8c9239-9b59-473a-a229-d8a9693be270",
          },
        ],
        "default_category_uuid": "0c8c9239-9b59-473a-a229-d8a9693be270",
        "operand": "@child.status",
        "type": "switch",
      },
      "uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
    },
    "ui": Object {
      "config": Object {},
      "position": null,
      "type": "split_by_subflow",
    },
  },
]
`;

exports[`SubflowRouterForm updates creates its own action uuid if necessary 1`] = `
Array [
  Object {
    "inboundConnections": Object {},
    "node": Object {
      "actions": Array [
        Object {
          "flow": Object {
            "name": "Favorite Color",
            "type": "messaging",
            "uuid": "9a93ede6-078f-44c9-ad0a-133793be5d56",
          },
          "type": "enter_flow",
          "uuid": "132de855-4042-4dc1-a18f-cc2e6a8f790a",
        },
      ],
      "exits": Array [
        Object {
          "destination_uuid": null,
          "uuid": "2dc85899-0577-4f1b-a620-f12094e34b5e",
        },
        Object {
          "destination_uuid": null,
          "uuid": "e67ce051-f3e4-4c42-807b-e393df5fa382",
        },
      ],
      "router": Object {
        "cases": Array [
          Object {
            "arguments": Array [
              "completed",
            ],
            "category_uuid": "42ebf1d0-a43e-4ba6-9577-c863b20c88c4",
            "type": "has_only_text",
            "uuid": "70ac6ea6-803a-4c33-81c7-dc26803c313f",
          },
          Object {
            "arguments": Array [
              "expired",
            ],
            "category_uuid": "e0bc3776-792f-49d2-b4be-48be2cf8159a",
            "type": "has_only_text",
            "uuid": "6c22884b-0e35-4206-982e-18320691eda9",
          },
        ],
        "categories": Array [
          Object {
            "exit_uuid": "2dc85899-0577-4f1b-a620-f12094e34b5e",
            "name": "Complete",
            "uuid": "42ebf1d0-a43e-4ba6-9577-c863b20c88c4",
          },
          Object {
            "exit_uuid": "e67ce051-f3e4-4c42-807b-e393df5fa382",
            "name": "Expired",
            "uuid": "e0bc3776-792f-49d2-b4be-48be2cf8159a",
          },
        ],
        "default_category_uuid": "e0bc3776-792f-49d2-b4be-48be2cf8159a",
        "operand": "@child.status",
        "type": "switch",
      },
      "uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
    },
    "ui": Object {
      "config": Object {},
      "position": null,
      "type": "split_by_subflow",
    },
  },
]
`;

exports[`SubflowRouterForm updates creates its own action uuid if necessary 2`] = `
Array [
  Object {
    "inboundConnections": Object {},
    "node": Object {
      "actions": Array [
        Object {
          "flow": Object {
            "name": "Favorite Sound",
            "type": "voice",
            "uuid": "80b0a850-6973-11ed-9c42-ad7d71773e93",
          },
          "type": "enter_flow",
          "uuid": "6b59e863-2e14-442b-9987-aaa284871b8e",
        },
      ],
      "exits": Array [
        Object {
          "destination_uuid": null,
          "uuid": "00dd2423-0a34-4ee2-b97d-839fb2d593ef",
        },
      ],
      "router": Object {
        "cases": Array [
          Object {
            "arguments": Array [
              "completed",
            ],
            "category_uuid": "c5722340-f7e1-4d99-ac5e-3a6e59845111",
            "type": "has_only_text",
            "uuid": "f05fa107-03db-4315-a211-53da63603e6f",
          },
        ],
        "categories": Array [
          Object {
            "exit_uuid": "00dd2423-0a34-4ee2-b97d-839fb2d593ef",
            "name": "Complete",
            "uuid": "c5722340-f7e1-4d99-ac5e-3a6e59845111",
          },
        ],
        "default_category_uuid": "c5722340-f7e1-4d99-ac5e-3a6e59845111",
        "operand": "@child.status",
        "type": "switch",
      },
      "uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
    },
    "ui": Object {
      "config": Object {},
      "position": null,
      "type": "split_by_subflow",
    },
  },
]
`;

exports[`SubflowRouterForm updates should update and save 1`] = `
Object {
  "flow": Object {
    "validationFailures": Array [],
    "value": Object {
      "name": "Favorite Color",
      "parent_refs": Array [
        "colors",
      ],
      "type": "messaging",
      "uuid": "9a93ede6-078f-44c9-ad0a-133793be5d56",
    },
  },
  "params": Object {
    "colors": Object {
      "value": "",
    },
  },
  "valid": true,
}
`;

exports[`SubflowRouterForm updates should update and save 2`] = `
Array [
  Object {
    "inboundConnections": Object {},
    "node": Object {
      "actions": Array [
        Object {
          "flow": Object {
            "name": "Favorite Color",
            "type": "messaging",
            "uuid": "9a93ede6-078f-44c9-ad0a-133793be5d56",
          },
          "type": "enter_flow",
          "uuid": "d4a3a01c-1dee-4324-b107-4ac7a21d836f",
        },
      ],
      "exits": Array [
        Object {
          "destination_uuid": null,
          "uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
        },
        Object {
          "destination_uuid": null,
          "uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
        },
      ],
      "router": Object {
        "cases": Array [
          Object {
            "arguments": Array [
              "C",
            ],
            "category_uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
            "type": "has_run_status",
            "uuid": "132de855-4042-4dc1-a18f-cc2e6a8f790a",
          },
          Object {
            "arguments": Array [
              "E",
            ],
            "category_uuid": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
            "type": "has_run_status",
            "uuid": "2dc85899-0577-4f1b-a620-f12094e34b5e",
          },
        ],
        "categories": Array [
          Object {
            "exit_uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
            "name": "Complete",
            "uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
          },
          Object {
            "exit_uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
            "name": "Expired",
            "uuid": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
          },
        ],
        "default_category_uuid": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
        "operand": "@child.status",
        "type": "switch",
      },
      "uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
    },
    "ui": Object {
      "config": Object {},
      "position": null,
      "type": "split_by_subflow",
    },
  },
]
`;

exports[`SubflowRouterForm updates should update and save 3`] = `
Object {
  "flow": Object {
    "validationFailures": Array [],
    "value": Object {
      "name": "Favorite Sound",
      "parent_refs": Array [
        "sounds",
      ],
      "type": "voice",
      "uuid": "80b0a850-6973-11ed-9c42-ad7d71773e93",
    },
  },
  "params": Object {
    "sounds": Object {
      "value": "",
    },
  },
  "valid": true,
}
`;

exports[`SubflowRouterForm updates should update and save 4`] = `
Array [
  Object {
    "inboundConnections": Object {},
    "node": Object {
      "actions": Array [
        Object {
          "flow": Object {
            "name": "Favorite Sound",
            "type": "voice",
            "uuid": "80b0a850-6973-11ed-9c42-ad7d71773e93",
          },
          "type": "enter_flow",
          "uuid": "2d1284d0-6972-11ed-9ca7-8f291517c832",
        },
      ],
      "exits": Array [
        Object {
          "destination_uuid": null,
          "uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
        },
      ],
      "router": Object {
        "cases": Array [
          Object {
            "arguments": Array [
              "C",
            ],
            "category_uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
            "type": "has_run_status",
            "uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
          },
        ],
        "categories": Array [
          Object {
            "exit_uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
            "name": "Complete",
            "uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
          },
        ],
        "default_category_uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
        "operand": "@child.status",
        "type": "switch",
      },
      "uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
    },
    "ui": Object {
      "config": Object {},
      "position": null,
      "type": "split_by_subflow",
    },
  },
]
`;
